### PR TITLE
[OSD-11798] Migrate certman-operator to operator-sdk 1.21

### PR DESCRIPTION
**Pull request by ravitri**
This PR regards the operator-sdk migration work as per [OSD-11798](https://issues.redhat.com/browse/OSD-11798).

Golang - 1.17
Openshift API - 4.11 release
Kubernetes - 1.24
Controller-Runtime - 0.12.1
Operator-SDK - 1.21

### CODE REVIEW

**Comment by ravitri** [comment]
/hold
For development and testing purposes, temporarily pointed to github.com/ravitri/aws-account-operator/api instead of github.com/openshift/aws-account-operator/api. Once PR https://github.com/openshift/aws-account-operator/pull/725 is merged, will point to the openshift repository. Till then, keeping the PR on hold for review.

File: controllers/certificaterequest_controller.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.2/pkg/reconcile
+func (r *CertificateRequestReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	// TODO(user): your logic here
+
+	return ctrl.Result{}, nil
+}
```
Comment (yithian) [inline controllers/certificaterequest_controller.go:55]: this doesn't look right to me :laughing:

**Review by yithian — CHANGES_REQUESTED** [review]

File: controllers/certificaterequest/issue_certificate_test.go [hunk 2]
```diff
@@ -83,7 +83,7 @@ func TestIssueCertificate(t *testing.T) {
 		t.Run(test.Name, func(t *testing.T) {
 			t.Helper()
 
-			nullLogger := logrTesting.NullLogger{}
+			nullLogger := logrTesting.Discard()
```
[the remark this answers is not in the document]
Comment (ravitri) [reply controllers/certificaterequest/issue_certificate_test.go:86]: Yup, agreed. Thanks Alex!

File: controllers/certificaterequest_controller.go (hunk 1 above)
Comment (ravitri) [reply controllers/certificaterequest_controller.go:55]: Oh I missed removing this! Thanks to catch it :+1: 

**Background**: What happened here is that when new controller is initialised using `operator-sdk` command, the new file is directly created under `./controllers` directory in the new version. However, as part of the migration I wanted to maintain directory structure so had manually copied controller directories from `./pkg/controller/` in old version. Missed removing this file later on. For the actual controller can refer to this [diff](https://github.com/openshift/certman-operator/pull/225/files#diff-0285b65fbbf63a5eb3c040c52cfac34e69791b73cfc5e3adeb59954b6432a7ad).

File: go.mod [hunk 3]
```diff
@@ -1,66 +1,109 @@
...
-	github.com/spf13/pflag v1.0.5
-	github.com/stretchr/testify v1.6.1
-	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
-	google.golang.org/api v0.25.0
-	k8s.io/api v0.19.0
-	k8s.io/apimachinery v0.19.0
-	k8s.io/client-go v12.0.0+incompatible
-	k8s.io/kube-openapi v0.0.0-20200805222855-6aeccd4b50c6
-	sigs.k8s.io/controller-runtime v0.6.2
+	github.com/ravitri/aws-account-operator/api v0.0.0-20220902055620-a0d038965be6
```
Comment (ravitri) [inline go.mod:21]: **NOTE**: This is temporary till https://github.com/openshift/aws-account-operator/pull/725 is merged. Once it is merged, migrated certman-operator to be tested again with any rebase changes of migrated AAO. If all works, will point to upstream AAO repository then.

File: controllers/certificaterequest/issue_certificate_test.go [hunk 4]
```diff
@@ -23,14 +23,14 @@ import (
 	"testing"
 
 	"github.com/eggsampler/acme"
-	logrTesting "github.com/go-logr/logr/testing"
+	logrTesting "github.com/go-logr/logr"
```
[the remark this answers is not in the document]
Comment (ravitri) [reply controllers/certificaterequest/issue_certificate_test.go]: Thanks Alex! Resolved with [commit](https://github.com/openshift/certman-operator/pull/225/commits/0fd636976046016ef946878a8b2b47e778d23ccc).

File: controllers/certificaterequest/renew_certificate_test.go [hunk 5]
```diff
@@ -19,10 +19,10 @@ package certificaterequest
 import (
 	"testing"
 
-	logrTesting "github.com/go-logr/logr/testing"
+	logrTesting "github.com/go-logr/logr"
```
[the remark this answers is not in the document]
Comment (ravitri) [reply controllers/certificaterequest/renew_certificate_test.go]: Thanks Alex! Resolved with [commit](https://github.com/openshift/certman-operator/pull/225/commits/0fd636976046016ef946878a8b2b47e778d23ccc).

**Comment by ravitri** [comment]
Thanks Alex :+1: Have added the fixes based on feedback.

File: controllers/certificaterequest_controller.go (hunk 1 above)
Comment (yithian) [reply controllers/certificaterequest_controller.go]: that makes sense, thanks for clarifying!

File: go.mod (hunk 3 above)
Comment (yithian) [reply go.mod:21]: i don't have any significant concerns once this is resolved

thanks for the big PR!

File: main.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	namespace, err := k8sutil.GetWatchNamespace()
+	if err != nil {
+		log.Error(err, "Failed to get watch namespace")
+		os.Exit(1)
+	}
+
+	if err := routev1.AddToScheme(clientgoscheme.Scheme); err != nil {
+		log.Error(err, "unable to add routev1 scheme")
+		os.Exit(1)
+	}
```
Comment (mrbarge) [inline main.go:127]: A minor nit, but if `operator-custom-metrics` dependency is bumped up to at least `v0.5.1-0.20220802235640-dc76a1f15ee8` or greater then it should prevent the need for this. (I'm assuming this was added to allow the operator to create the metrics route successfully)

File: go.mod (hunk 3 above)
Comment (ravitri) [reply go.mod]: This has been updated now as the AAO PR is merged.

File: main.go (hunk 6 above)
Comment (ravitri) [reply main.go]: Thanks a lot Matt! Yup, have updated the dependency and tested it as well. Seemed to work fine and certman-operator didn't complain of the route API being unknown.

**Comment by ravitri** [comment]
Based on the testing and comments so far, am ok to unhold this PR (unless anyone has any other concerns)
/unhold

**Comment by mrbarge** [comment]
I've completed testing this operator as part of OSD-12975. A full write-up of testing is available in that ticket, but the tl;dr is that no issues stemming from this PR were identified and the operator is behaving correctly.

**Pull request merged**